python/lsst/resources/http.py

```diff
@@ -531,49 +580,269 @@ def transfer_from(
             )
-            if resp.is_redirect or resp.is_permanent_redirect:
-                final_url = resp.headers["Location"]
-                log.debug("PUT request to %s redirected to %s", self.geturl(), final_url)
+        return self._send_webdav_request("PROPFIND", headers=headers, body=body)
+
+    def _mkcol(self) -> None:
+        """Send a MKCOL webDAV request to create a collection. The collection
+        may already exist.
+        """
+        resp = self._send_webdav_request("MKCOL")
+        if resp.status_code == requests.codes.created:  # 201
+            return
+
+        if resp.status_code == requests.codes.method_not_allowed:  # 405
+            # The remote directory already exists
+            log.debug("Can not create directory: %s may already exist: skipping.", self.geturl())
+        else:
+            raise ValueError(f"Can not create directory {self}, status code: {resp.status_code}")
+
+    def _delete(self) -> None:
+        """Send a DELETE webDAV request for this resource."""
+        # TODO: should we first check that the resource is not a directory?
+        # TODO: we should remove Depth header which should not be used for
+        # directories and systematically check if the return code is
+        # multistatus, in which case we need to look for errors in the
+        # response.
+        resp = self._send_webdav_request("DELETE", headers={"Depth": "0"})
+        if resp.status_code in (requests.codes.ok, requests.codes.accepted, requests.codes.no_content):
+            return
+        elif resp.status_code == requests.codes.not_found:
+            raise FileNotFoundError(f"Resource {self} does not exist, status code: {resp.status_code}")
+        else:
+            raise ValueError(f"Unable to delete resource {self}; status code: {resp.status_code}")
 
-        # Send data to its final destination.
-        resp = self.put_session.put(final_url, data=data, timeout=TIMEOUT)
-        if resp.status_code not in [200, 201, 202, 204]:
+    def _copy(self, src: HttpResourcePath) -> None:
+        """Send a COPY webDAV request to replace the contents of this resource
+        (if any) with the contents of another resource located in the same
+        server.
+
+        Parameters
+        ----------
+        src : `HttpResourcePath`
+            The source of the contents to copy to `self`.
+        """
+        # Neither dCache nor XrootD currently implement the COPY
+        # webDAV method as documented in
+        #    http://www.webdav.org/specs/rfc4918.html#METHOD_COPY
+        # (See issues DM-37603 and DM-37651 for details)
+        #
+        # For the time being, we use a temporary local file to
+        # perform the copy client side.
+        # TODO: when those 2 issues above are solved remove the 3 lines below.
+        must_use_local = True
+        if must_use_local:
+            return self._copy_via_local(src)
+
+        headers = {"Destination": self.geturl()}
```

Comment:
This code from here to the end of the file is the same as the `_move` code below but with a different verb. Since this code is not used we could delete it. We could also change `def _move` to be `def _move_or_copy` and have it take an `action="MOVE"` optional parameter.

Reply:
Fixed. Introduced helper method `copy_or_move()`.